File: templates/next-supabase-tambo-notes/nexus_backend/app/services/event_service.py
"""
Event Service for real-time event publishing via SSE
"""
from typing import Dict, Any, Optional, Callable, List
from uuid import UUID
import asyncio
import json
import structlog
from datetime import datetime

logger = structlog.get_logger()


class EventService:
    """In-memory event service for SSE broadcasting"""
    
    _instance: Optional['EventService'] = None
    _subscribers: Dict[str, List[Callable]] = {}  # user_id -> list of callbacks
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance
    
    def subscribe(self, user_id: UUID, callback: Callable) -> None:
        """Subscribe a user to events"""
        user_id_str = str(user_id)
        if user_id_str not in self._subscribers:
            self._subscribers[user_id_str] = []
        self._subscribers[user_id_str].append(callback)
        logger.info("User subscribed to events", user_id=user_id_str, subscribers=len(self._subscribers[user_id_str]))
    
    def unsubscribe(self, user_id: UUID, callback: Callable) -> None:
        """Unsubscribe a user from events"""
        user_id_str = str(user_id)
        if user_id_str in self._subscribers:
            try:
                self._subscribers[user_id_str].remove(callback)
                if not self._subscribers[user_id_str]:
                    del self._subscribers[user_id_str]
                logger.info("User unsubscribed from events", user_id=user_id_str)
            except ValueError:
                pass
    
    async def publish(
        self,
        user_id: UUID,
        event_type: str,
        data: Dict[str, Any],
        event_id: Optional[str] = None
    ) -> None:
        """
        Publish an event to a specific user.
        
        Args:
            user_id: Target user ID
            event_type: Event type (e.g., 'task_created', 'status_changed')
            data: Event data payload
            event_id: Optional event ID for SSE
        """
        user_id_str = str(user_id)
        
        event = {
            "id": event_id or f"{int(datetime.utcnow().timestamp() * 1000)}",
            "type": event_type,
            "data": json.dumps(data),
            "timestamp": datetime.utcnow().isoformat(),
        }
        
        if user_id_str in self._subscribers:
            # Send to all subscribers for this user
            for callback in self._subscribers[user_id_str]:
                try:
                    await callback(event)
                except Exception as e:
                    logger.error("Error sending event to subscriber", error=str(e), user_id=user_id_str)
        
        logger.debug("Event published", user_id=user_id_str, event_type=event_type)
    
    async def publish_broadcast(
        self,
        event_type: str,
        data: Dict[str, Any],
        event_id: Optional[str] = None
    ) -> None:
        """Broadcast event to all subscribers"""
        for user_id_str in list(self._subscribers.keys()):
            user_id = UUID(user_id_str)
            await self.publish(user_id, event_type, data, event_id)
    
    def get_subscriber_count(self, user_id: Optional[UUID] = None) -> int:
        """Get number of subscribers (for a user or total)"""
        if user_id:
            return len(self._subscribers.get(str(user_id), []))
        return sum(len(callbacks) for callbacks in self._subscribers.values())


# Global instance
event_service = EventService()

File: templates/next-supabase-tambo-notes/src/lib/nexus/event-service.ts
/**
 * Event Service
 * In-memory event service for SSE real-time updates
 */

// UUID type - using string for compatibility

export interface NexusEvent {
  type: string
  data: Record<string, unknown>
  timestamp: string
}

export interface EventQueue {
  queue: NexusEvent[]
  maxSize: number
  ttl: number // milliseconds
}

export class EventService {
  private static subscribers: Map<string, Set<AsyncQueue<NexusEvent>>> = new Map()
  private static eventQueues: Map<string, EventQueue> = new Map()
  private static readonly MAX_QUEUE_SIZE = 100
  private static readonly EVENT_TTL = 60000 // 60 seconds
  private static readonly MAX_CONNECTIONS_PER_USER = 5

  /**
   * Subscribe to events for a user
   * Returns an async queue that will receive events
   */
  static subscribe(userId: string): AsyncQueue<NexusEvent> {
    // Check connection limit
    const userSubscribers = this.subscribers.get(userId) || new Set()
    if (userSubscribers.size >= this.MAX_CONNECTIONS_PER_USER) {
      throw new Error('Maximum connections per user exceeded')
    }

    const queue = new AsyncQueue<NexusEvent>()
    
    if (!this.subscribers.has(userId)) {
      this.subscribers.set(userId, new Set())
    }
    
    this.subscribers.get(userId)!.add(queue)

    // Initialize event queue for user if not exists
    if (!this.eventQueues.has(userId)) {
      this.eventQueues.set(userId, {
        queue: [],
        maxSize: this.MAX_QUEUE_SIZE,
        ttl: this.EVENT_TTL,
      })
    }

    console.log(`[EventService] User ${userId} subscribed. Total subscribers: ${userSubscribers.size + 1}`)
    
    return queue
  }

  /**
   * Unsubscribe a user's queue
   */
  static unsubscribe(userId: string, queue: AsyncQueue<NexusEvent>): void {
    const userSubscribers = this.subscribers.get(userId)
    if (userSubscribers) {
      userSubscribers.delete(queue)
      if (userSubscribers.size === 0) {
        this.subscribers.delete(userId)
        this.eventQueues.delete(userId)
      }
      console.log(`[EventService] User ${userId} unsubscribed. Remaining subscribers: ${userSubscribers.size}`)
    }
  }

  /**
   * Publish an event to all subscribers of a user
   */
  static publish(userId: string, event: Omit<NexusEvent, 'timestamp'>): void {
    const fullEvent: NexusEvent = {
      ...event,
      timestamp: new Date().toISOString(),
    }

    // Add to event queue for this user
    const eventQueue = this.eventQueues.get(userId)
    if (eventQueue) {
      // Clean up old events
      const now = Date.now()
      eventQueue.queue = eventQueue.queue.filter(
        (e) => now - new Date(e.timestamp).getTime() < eventQueue.ttl
      )

      // Add new event
      eventQueue.queue.push(fullEvent)

      // Enforce max size
      if (eventQueue.queue.length > eventQueue.maxSize) {
        eventQueue.queue.shift()
      }
    }

    // Send to all active subscribers
    const userSubscribers = this.subscribers.get(userId)
    if (userSubscribers) {
      let sentCount = 0
      for (const queue of userSubscribers) {
        try {
          queue.enqueue(fullEvent)
          sentCount++
        } catch (error) {
          console.error(`[EventService] Failed to send event to queue:`, error)
          // Remove dead queue
          userSubscribers.delete(queue)
        }
      }
      console.log(`[EventService] Published event ${event.type} to ${sentCount} subscribers for user ${userId}`)
    }
  }

  /**
   * Get pending events for a user (for catch-up after reconnection)
   */
  static getPendingEvents(userId: string, since?: string): NexusEvent[] {
    const eventQueue = this.eventQueues.get(userId)
    if (!eventQueue) {
      return []
    }

    if (since) {
      return eventQueue.queue.filter((e) => e.timestamp > since)
    }

    return [...eventQueue.queue]
  }

  /**
   * Get subscriber count for a user (for monitoring)
   */
  static getSubscriberCount(userId: string): number {
    return this.subscribers.get(userId)?.size || 0
  }
}

/**
 * Simple async queue implementation for event streaming
 */
export class AsyncQueue<T> {
  private items: T[] = []
  private waiters: Array<(item: T) => void> = []

  async dequeue(): Promise<T> {
    if (this.items.length > 0) {
      return this.items.shift()!
    }

    return new Promise<T>((resolve) => {
      this.waiters.push(resolve)
    })
  }

  enqueue(item: T): void {
    if (this.waiters.length > 0) {
      const waiter = this.waiters.shift()!
      waiter(item)
    } else {
      this.items.push(item)
    }
  }

  isEmpty(): boolean {
    return this.items.length === 0 && this.waiters.length === 0
  }

  clear(): void {
    this.items = []
    this.waiters = []
  }
}
